bin/compile

```diff
@@ -46,3 +46,16 @@ topic "Creating chromedriver export scripts"
 mkdir -p $BUILD_DIR/.profile.d
 echo "export PATH=\$PATH:\$HOME/.chromedriver/bin" >> $BUILD_DIR/.profile.d/chromedriver.sh
 indent "Created"
+
+if [ -x "$BUILD_DIR/.apt/usr/bin/google-chrome" ]; then
```

Comment:
I have just gotten the feeling this is a big problem than I expected. :sob:

`$GOOGLE_CHROME_SHIM` is undefined while buildpack is compiling.

```
remote: -----> Creating chromedriver export scripts...
remote: /app/tmp/buildpacks/5125a5e2d165927f5d0e43b76f51521104bbb5800661036228c9e9d66d8e61aba110b386a45b029db74de0672b69f019590fc046951f76ab671cfe64f1cb811c/bin/compile: line
54: GOOGLE_CHROME_SHIM: unbound variable
```

So, I changed to the below:

```sh
if [ -r "$BUILD_DIR/.profile.d/010_google-chrome.sh" ]; then # <= Added
  source $BUILD_DIR/.profile.d/010_google-chrome.sh          # <= Added
fi                                                           # <= Added

if [ -x "$GOOGLE_CHROME_SHIM" ]; then
  topic "Validating chromedriver"
  $BIN_DIR/chromedriver --version | indent
  echo "Validated" | indent
else
  cat <<WARNING
You should set heroku-buildpack-google-chrome before heroku-buildpack-chromedriver. Then you can validate the chromedriver binary.

1. https://github.com/heroku/heroku-buildpack-google-chrome.git
2. https://github.com/heroku/heroku-buildpack-chromedriver.git
WARNING
fi
```

The content of `010_google-chrome.sh` is below:

```sh
export GOOGLE_CHROME_BIN="$HOME/.apt/opt/google/chrome/chrome"
export GOOGLE_CHROME_SHIM="$HOME/.apt/usr/bin/google-chrome-stable"
```

Unfortunately, `$GOOGLE_CHROME_SHIM` was not exist. Because `$HOME` was `/tmp/build_91ab22e41b0f330e61c1001c1edde73f` when compiling buildpack.

I feel that the dependency of chromedriver-buildpack and chrome-buildpack is strong.

What do you think about this?

Reply:
I got the feeling that chromedriver-buildpack should be install `libnss3.so`, `libnssutil3.so` and `libnspr4.so`.

```
$ heroku run bash
~ $ which chromedriver
/app/.chromedriver/bin/chromedriver
~ $ ldd /app/.chromedriver/bin/chromedriver
        linux-vdso.so.1 (0x00007fff6e976000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007ff303955000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ff303736000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007ff30352e000)
        libglib-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007ff303217000)
        libnss3.so => not found
        libnssutil3.so => not found
        libnspr4.so => not found
        libX11.so.6 => /usr/lib/x86_64-linux-gnu/libX11.so.6 (0x00007ff302edf000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ff302b41000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ff302929000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ff302538000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ff3045bd000)
        libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007ff3022c6000)
        libxcb.so.1 => /usr/lib/x86_64-linux-gnu/libxcb.so.1 (0x00007ff30209e000)
        libXau.so.6 => /usr/lib/x86_64-linux-gnu/libXau.so.6 (0x00007ff301e9a000)
        libXdmcp.so.6 => /usr/lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007ff301c94000)
        libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007ff301a7f000)
```

So, I made a prototype referring to [heroku-buildpack-google-chrome](https://github.com/heroku/heroku-buildpack-google-chrome/blob/737a426cec8327daa80a2a79a0f695b45f9bc3d1/bin/compile#L60-L139).
https://github.com/heroku/heroku-buildpack-chromedriver/compare/master...masutaka:add-validator-v3

There are many problems. :confused:

* Big change
* The compile time increase
* Maybe need to drop `cedar-14` stack
* How to create the setting file of `LD_LIBRARY_PATH`

Reply:
Yike, yeah that exploded more than I expected. I think the buildpack to buildpack reference is too great if we start trying to look for implementation details like your first comment.

I do think that updating this buildpack to be completely functional without chrome would be ideal. This could open up additional use cases of connecting from heroku to a remote chrome instance. We would still need to support  cedar-14 until sometime next year at the earliest. The `heroku-buildpack-google-chrome` supports all 3 stacks today. The dependencies are likely a bit different, but still should be possible. The `LD_LIBRARY_PATH` is also a concern, as you say. We might need to duplicate the logic [here](https://github.com/heroku/heroku-buildpack-google-chrome/blob/737a426cec8327daa80a2a79a0f695b45f9bc3d1/bin/compile#L161). I think that thinking about these 2 buildpacks as not related will give the best flexibility.

Even thinking about it now I can see where the new warning wouldn't make sense to users driving a chrome instance off-dyno once they've met all the requirements to run the driver itself.

Reply:
I have understood that it is more complicated than expected. :sob:

I will close this PR later. I hope we can add dependencies to the buildpack!

By the way, I believe https://github.com/heroku/heroku-buildpack-google-chrome/pull/72 is natural PR. Please review if possible.